plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineRunner.java

```diff
@@ -0,0 +1,114 @@
+
+import com.google.cloud.tools.eclipse.dataflow.core.project.MajorVersion;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.SetMultimap;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * All available {@code PipelineRunner} implementations.
+ */
+public enum PipelineRunner {
+  DIRECT_PIPELINE_RUNNER(
+      "DirectPipelineRunner",
+      "com.google.cloud.dataflow.sdk.options.DirectPipelineOptions",
+      "Runs the pipeline on the local machine.",
+      EnumSet.of(MajorVersion.ONE)),
+  DIRECT_RUNNER(
+      "DirectRunner",
+      "org.apache.beam.runners.direct.DirectOptions",
+      "Runs the pipeline on the local machine.",
+      EnumSet.of(MajorVersion.QUALIFIED_TWO, MajorVersion.TWO, MajorVersion.THREE_PLUS)),
+  DATAFLOW_PIPELINE_RUNNER(
+      "DataflowPipelineRunner",
+      "com.google.cloud.dataflow.sdk.options.DataflowPipelineOptions",
+      "Runs the pipeline remotely on the Dataflow service and detach.",
+      EnumSet.of(MajorVersion.ONE)),
+  DATAFLOW_RUNNER(
+      "DataflowRunner",
+      "org.apache.beam.runners.dataflow.options.DataflowPipelineOptions",
+      "Runs the pipeline remotely on the Dataflow service and detach.",
+      EnumSet.of(MajorVersion.QUALIFIED_TWO, MajorVersion.TWO, MajorVersion.THREE_PLUS)),
+  BLOCKING_DATAFLOW_PIPELINE_RUNNER(
+      "BlockingDataflowPipelineRunner",
+      "com.google.cloud.dataflow.sdk.options.BlockingDataflowPipelineOptions",
+      "Runs the pipeline remotely on the Dataflow service and wait for it to complete.",
+      EnumSet.of(MajorVersion.ONE));
+
+  private static final Map<String, PipelineRunner> RUNNERS_BY_NAME;
+  private static final SetMultimap<MajorVersion, PipelineRunner> RUNNERS_IN_VERSION;
+
+  static {
+    ImmutableMap.Builder<String, PipelineRunner> runnersByName = ImmutableMap.builder();
+    ImmutableSetMultimap.Builder<MajorVersion, PipelineRunner> runnersInVersion =
+        ImmutableSetMultimap.builder();
+    for (PipelineRunner runner : PipelineRunner.values()) {
+      runnersByName.put(runner.getRunnerName(), runner);
+      for (MajorVersion supported : runner.supportedVersions) {
+        runnersInVersion.put(supported, runner);
+      }
+    }
+    RUNNERS_BY_NAME = runnersByName.build();
+    RUNNERS_IN_VERSION = runnersInVersion.build();
+  }
+
+  public static PipelineRunner fromRunnerName(String runnerName) {
+    return RUNNERS_BY_NAME.get(runnerName);
+  }
+
+  /**
+   * Return the set of all Pipeline Runners in the provided {@link MajorVersion}. Should be used
+   * instead of {@link #values()}.
+   */
+  public static Set<PipelineRunner> inMajorVersion(MajorVersion majorVersion) {
+    return RUNNERS_IN_VERSION.get(majorVersion);
+  }
+
+  private final String runnerName;
+  private final String optionsName;
+  private final String description;
+  private final Set<MajorVersion> supportedVersions;
```

Comment:
We haven't really been using feature branches until now. Until other people start committing to your branch, rewriting history should not be a problem.